Python/chapter02/2.4 - Partition/miguel_2.4_soln.py

```diff
@@ -0,0 +1,241 @@
+            self.tail = self.head
+        else:
+            end = Node(d)
+            self.tail.next = end
+            self.tail = end
+        self.size += 1
+
+    def append_to_head(self, d: int) -> None:
+        new_head = Node(d)
+        new_head.next = self.head
+        self.head = new_head
+        self.size += 1
+
+    def get_node_at(self, index: int) -> Node:
+        if index < 0 or index >= self.size:
+            raise IndexError('list index out of range')
+        n = self.head
+        for i in range(self.size):
+            if i == index:
+                return n
+            n = n.next
+
+    def __repr__(self):
+        return self.__str__()
+
+    def __str__(self):
+        if self.head is None:
+            return '<empty>'
+        ll = []
+        n = self.head
+        while n.next is not None:
+            ll.append('{} -> '.format(n.data))
+            n = n.next
+        ll.append(str(n.data))
+        return ''.join(ll)
+
+    def __eq__(self, other: object):
+        if not isinstance(other, LinkedList):
+            return NotImplemented
+        a = self.head
+        b = other.head
+        while a is not None and b is not None:
+            if a.data != b.data:
+                return False
+            # otherwise, advance both pointers
+            a = a.next
+            b = b.next
+        return a is None and b is None
+
+
+def check_partitioned(ll: LinkedList, partition: int) -> bool:
+    """
+    Will check if a linked list is partitioned
+    around a value such that all values less
+    then the partition value come before all values
+    greater than or equal to the partition value
+    in the linked list.
+    Order doesn't matter, as long as the values
+    < partition come before values >= partition.
+    Runtime: O(n)
+    Space Complexity: O(1)
+    :param ll: an input linked list
+    :param partition: a number to partition around
+    :return:
+    """
+    if ll.size == 1:
+        return True
+    # counter for keeping track of how many times
+    # we change from < partition to >= partition
+    partition_change_count = 0
+    changed = False
+    n = ll.head
+    while n is not None:
+        if n.data < partition and changed:
+            changed = False
+            partition_change_count += 1
+        elif n.data >= partition and not changed:
+            changed = True
+            partition_change_count += 1
+        n = n.next
+    return partition_change_count <= 1
+
+
+def partition_ll(ll: LinkedList, partition: int) -> LinkedList:
+    """
+    This function will take in a linked list, a
+    partition value, and will partition a linked
+    list around a value x such that all nodes
+    less than x come before all nodes greater than
+    or equal to x.  If x is in the list, then
+    the value(s) of x only need to be after the
+    elements less than x. They do not need to appear
+    in between the left and right partitions.
+    Runtime: O(n)
+    Space Complexity: O(n)
```

Comment:
Is this optimal?

Reply:
Not optimal because we can instead swap in place, leading to O(1) space complexity, correct?

Reply:
correct. you can also move the nodes to two temporary linkedlists and then splice them together.

neither alternative would worsen the runtime complexity.

Reply:
Wouldn't creating two temporary linked lists result in O(n) space still?  Also, since a new linked list is the output, wouldn't that mean O(n) space either way?

Reply:
not if you move the nodes from one list to another. the list itself is just two pointers and a size.

Reply:
Ah, I see.